After the node is upgraded to the latest version, data cannot be synchronized

v1.2.13 => v1.3.5-hotfix2

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1335195]

goroutine 2026591 [running]:
github.com/spacemeshos/go-spacemesh/blocks.(*Generator).Stop(0xc07daa4800)
	/opt/smhmain/core/go-spacemesh/blocks/generator.go:124 +0x35
github.com/spacemeshos/go-spacemesh/node.(*App).stopServices(0xc000408380, {0x1e29848?, 0xc000369810?})
	/opt/smhmain/core/go-spacemesh/node/node.go:1468 +0x385
github.com/spacemeshos/go-spacemesh/node.(*App).Cleanup(0xc000408380, {0x1e29848, 0xc000369810})
	/opt/smhmain/core/go-spacemesh/node/node.go:504 +0x6d
github.com/spacemeshos/go-spacemesh/node.GetCommand.func1.1.1()
	/opt/smhmain/core/go-spacemesh/node/node.go:191 +0x2d
created by github.com/spacemeshos/go-spacemesh/node.GetCommand.func1.1 in goroutine 1
	/opt/smhmain/core/go-spacemesh/node/node.go:190 +0x412

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,13 @@
 
 See [RELEASE](./RELEASE.md) for workflow instructions.
 
+## UNRELEASED
+
+### Improvements
+
+* [#5441](https://github.com/spacemeshos/go-spacemesh/pull/5441)
+  Fix possible nil-pointer dereference in blocks.Generator.
+
 ## Release v1.3.7
 
 ### Improvements

--- a/README.md
+++ b/README.md
@@ -444,6 +444,7 @@ as on UNIX-based systems.
 **WARNING! THIS IS EXPERIMENTAL FUNCTIONALITY, USE WITH CARE!**
 
 In order to make the p2p network more decentralized, the following options are provided:
+
 - `"enable-routing-discovery": true`: enables routing discovery for finding new peers, including those behind NAT, ans
   also for discovering relay nodes which are used for NAT hole punching. Note that hole punching can be done when both
   ends of the connection are behind an endpoint-independent ("cone") NAT.
@@ -457,6 +458,7 @@ In order to make the p2p network more decentralized, the following options are p
 - `"ping-peers": ["p2p_id_1", "p2p_id_2", ...]` runs P2P ping against the specified peers, logging the results.
 
 For the purpose of debugging P2P connectivity issues, the following command can also be used:
+
 ```console
 $ grpcurl -plaintext 127.0.0.1:9093 spacemesh.v1.DebugService.NetworkInfo
 {

--- a/blocks/generator.go
+++ b/blocks/generator.go
@@ -121,6 +121,9 @@ func (g *Generator) Start(ctx context.Context) {
 
 // Stop stops listening to hare output.
 func (g *Generator) Stop() {
+	if g.stop == nil {
+		return
+	}
 	g.stop()
 	err := g.eg.Wait()
 	if err != nil && !errors.Is(err, context.Canceled) {

--- a/blocks/generator_test.go
+++ b/blocks/generator_test.go
@@ -259,6 +259,11 @@ func Test_StartStop(t *testing.T) {
 	tg.Stop()
 }
 
+func Test_StopBeforeStart(t *testing.T) {
+	tg := createTestGenerator(t)
+	tg.Stop()
+}
+
 func genData(
 	t *testing.T,
 	db *sql.Database,

